src/audio/module_adapter/module/volume/volume.c

```diff
@@ -361,12 +362,33 @@ static void reset_state(struct vol_data *cd)
 }
 
 #if CONFIG_IPC_MAJOR_3
-static int init_volume(struct comp_dev *dev, struct comp_ipc_config *config, void *spec)
+static int volume_init(struct processing_module *mod)
 {
-	struct ipc_config_volume *vol = spec;
-	struct vol_data *cd = comp_get_drvdata(dev);
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct module_config *cfg = &md->cfg;
+	struct ipc_config_volume *vol = cfg->data;
+	struct vol_data *cd;
+	const size_t vol_size = sizeof(int32_t) * SOF_IPC_MAX_CHANNELS * 4;
 	int i;
 
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct vol_data));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
```

Comment:
I'd put this after the second allocation to avoid a dangling stale pointer in case the second allocation fails